// storybook-addon-themes/dist/vue.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var addons_1 = __importStar(require("@storybook/addons"));
var constants_1 = require("./constants");
var parameters_1 = __importDefault(require("./parameters"));
var store_1 = require("./store");
var shared_1 = require("./shared");
var vue_1 = require("./decorators/vue");
exports.store = new store_1.Store();
function wrapper(getStory, context, _a) {
    var parameters = _a.parameters;
    var config = shared_1.getConfig(parameters);
    var channel = addons_1.default.getChannel();
    channel.emit(constants_1.DECORATOR);
    return {
        components: { ThemeDecorator: vue_1.ThemeDecorator },
        template: "\n<theme-decorator :config=\"config\" :store=\"store\">\n  <story/>\n</theme-decorator>",
        data: function () {
            return { config: config, store: exports.store };
        },
    };
}
exports.withThemes = addons_1.makeDecorator(__assign(__assign({}, parameters_1.default), { wrapper: wrapper }));
if (module && module.hot && module.hot.decline) {
    module.hot.decline();
}
